test/hummingbot/connector/derivative/phemex_perpetual/test_phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,583 @@
+        }
+        return resp
+
+    @aioresponses()
+    def test_get_snapshot_exception_raised(self, mock_api):
+        url = web_utils.public_rest_url(CONSTANTS.SNAPSHOT_REST_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+        mock_api.get(regex_url, status=400, body=json.dumps(["ERROR"]))
+
+        with self.assertRaises(IOError) as context:
+            self.async_run_with_timeout(self.data_source._order_book_snapshot(trading_pair=self.trading_pair))
+
+        self.assertEqual(
+            'Error executing request GET https://testnet-api.phemex.com/md/v2/orderbook. HTTP status is 400. Error: ["ERROR"]',
+            str(context.exception),
+        )
+
+    @aioresponses()
+    def test_get_snapshot_successful(self, mock_api):
+        url = web_utils.public_rest_url(CONSTANTS.SNAPSHOT_REST_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+        mock_response = {
+            "error": None,
+            "id": 0,
+            "result": {
+                "orderbook_p": {
+                    "asks": [[87705000, 1000000], [87710000, 200000]],
+                    "bids": [[87700000, 2000000], [87695000, 200000]],
+                },
+                "depth": 30,
+                "sequence": 455476965,
+                "timestamp": 1583555482434235628,
+                "symbol": self.ex_trading_pair,
+                "type": "snapshot",
+            },
+        }
+        mock_api.get(regex_url, status=200, body=json.dumps(mock_response))
+
+        result: Dict[str, Any] = self.async_run_with_timeout(
+            self.data_source._request_order_book_snapshot(trading_pair=self.trading_pair)
+        )
+        self.assertEqual(mock_response, result)
+
+    @aioresponses()
+    def test_get_new_order_book(self, mock_api):
+        url = web_utils.public_rest_url(CONSTANTS.SNAPSHOT_REST_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+        mock_response = {
+            "error": None,
+            "id": 0,
+            "result": {
+                "orderbook_p": {
+                    "asks": [[87705000, 1000000], [87710000, 200000]],
+                    "bids": [[87700000, 2000000], [87695000, 200000]],
+                },
+                "depth": 30,
+                "sequence": 455476965,
+                "timestamp": 1583555482434235628,
+                "symbol": self.ex_trading_pair,
+                "type": "snapshot",
+            },
+        }
+        mock_api.get(regex_url, status=200, body=json.dumps(mock_response))
+        result = self.async_run_with_timeout(self.data_source.get_new_order_book(trading_pair=self.trading_pair))
+        self.assertIsInstance(result, OrderBook)
+        self.assertEqual(455476965, result.snapshot_uid)
+
+    @aioresponses()
+    def test_get_funding_info_from_exchange_successful(self, mock_api):
```

Comment:
It looks like this and the next test case are identical

Reply:
done

Reply:
Is a commit maybe missing? I still see `test_get_funding_info_from_exchange_successful` and `test_get_funding_info` being identical

Reply:
done